## 📍 Agenda 📍
<br>

- [Documentação oficial - Numpy](https://numpy.org/doc/stable/user/whatisnumpy.html)
- [Exercícios](#dois)

### Exercícios

1. Importe numpy como 'np' e imprima o número da versão instalada no ambiente (local ou colab).

In [40]:
import numpy as np
print(np.__version__)


1.23.4


2. Crie um *array* 1D com números de 0 a 9 e imprima o resultado.

In [6]:
vetor = np.array(range(1, 10, 1))
vetor

array([1, 2, 3, 4, 5, 6, 7, 8, 9])

In [43]:
arr_outro = np.arange(10)
arr_outro

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

3. Crie uma matriz **booleana** de 3x3 e preencha os seus valores com 'True'. Imprima o resultado.

**Dica:** Verifique o uso da função **full** na documentação do [numpy](https://numpy.org/doc/stable/user/whatisnumpy.html).

In [16]:
dimensoes = (3, 3)
matriz = vetor.reshape(dimensoes)
matriz

np.full((3, 3), True)


array([[ True,  True,  True],
       [ True,  True,  True],
       [ True,  True,  True]])

In [44]:
e = np.zeros((3,3))
boleana = e<1
boleana

array([[ True,  True,  True],
       [ True,  True,  True],
       [ True,  True,  True]])

In [45]:
e =np.zeros((3, 3))
vetor = e==0
vetor

array([[ True,  True,  True],
       [ True,  True,  True],
       [ True,  True,  True]])

4. Crie um *array* 1D como números inteiros (50 elementos) de 0 a 1000 e **extraia** apenas os números pares. Imprima o resultado.

In [47]:
arr = np.random.randint(1, 1000, 50)
inteiros = arr[arr % 2 == 0]
inteiros

array([278, 334, 910, 640, 250, 364, 708,  60,  20, 232, 176, 150, 716,
       574, 536, 250])

5. Usando o *array* criando no exercício anterior, **substitua** todos os números ímpares maiores que 50 por -1 no *array* original. 

In [53]:
arr = np.random.randint(1, 1000, 50)
arr[(arr > 50) & (arr % 2 != 0)] = - 1
arr

array([830,  -1, 674,  -1,  -1, 826, 892, 198,  -1,  -1, 914, 524,  11,
       894,  -1,  -1,  -1,  -1,  25, 182, 716,  -1,  -1, 708, 170,   4,
        72, 528, 892,  -1, 308, 816,  -1,  -1,  -1,  -1,  -1,  -1,  -1,
        -1, 382,  -1,  -1, 588,  -1,  -1,  52, 880,  -1,  -1])

6. Usando o *array* criando no exercício anterior, **substitua** todos os números ímpares com o valor -1 **sem** alterar o *array* original.

In [54]:
arr2 = np.where((arr>50) & (arr %2 == 0), arr, -1)
arr2

array([830,  -1, 674,  -1,  -1, 826, 892, 198,  -1,  -1, 914, 524,  -1,
       894,  -1,  -1,  -1,  -1,  -1, 182, 716,  -1,  -1, 708, 170,  -1,
        72, 528, 892,  -1, 308, 816,  -1,  -1,  -1,  -1,  -1,  -1,  -1,
        -1, 382,  -1,  -1, 588,  -1,  -1,  52, 880,  -1,  -1])

7. Crie um *array* 1D com 100 valores de ponto flutuante (**float**), converta-o em uma matriz 2D com 10 linhas e 10 colunas.

8. Crie duas matrizes 2D de inteiros aleatórios (de 0 a 100) com dimensões 10x10. **Empilhe-as horizontalmente**.

**Dica:** verique as funções **hstack** e **concatenate** na documentação do [numpy](https://numpy.org/doc/stable/user/whatisnumpy.html)

9. Crie duas matrizes 2D de inteiros aleatórios (de 0 a 100) com dimensões 10x10. **Empilhe-as verticalmente**.

Dica: verique as funções **vstack** e **concatenate** na documentação do [numpy](https://numpy.org/doc/stable/user/whatisnumpy.html)


Para ler arquivos em formato *txt* que estão em formato de "tabelas", podemos usar a função **genfromtxt()** do **numpy**. Esta função recebe como argumentos: **caminho do arquivo txt** (path local ou uma url); **delimiter** para separar cada valor seguindo algum critério; **dtype** para converter os valores recebidos em algum tipo de dado conhecido (exemplo: **float**) e; **usecols** em que é definido o cabeçalho das colunas lidas a partir do arquivo. 
<br>
<br>
Vejamos um exemplo de uso da função **genfromtxt** sobre a base de dados 
<a href="https://archive.ics.uci.edu/ml/datasets/iris">Iris</a>.

In [ ]:
import numpy as np
# Caminho da base de dados é uma url disponível na internet (visite o link abaixo
# pelo navegador para título de curiosidade :))
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'
# Lê os dados a partir do caminho passado e armazena em uma matriz 2D chamada iris_2d
iris_2d = np.genfromtxt(url, delimiter=',', dtype='float', usecols=[0,1,2,3])

In [ ]:
# matriz 2D criada com os dados da base Iris
iris_2d

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.7, 1.5, 0.2],
       [4.8, 3.4, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.3, 3. , 1.1, 0.1],
       [5.8, 4. , 1.2, 0.2],
       [5.7, 4.4, 1.5, 0.4],
       [5.4, 3.9, 1.3, 0.4],
       [5.1, 3.5, 1.4, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [5.1, 3.8, 1.5, 0.3],
       [5.4, 3.4, 1.7, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [4.6, 3.6, 1. , 0.2],
       [5.1, 3.3, 1.7, 0.5],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.2, 3.5, 1.5, 0.2],
       [5.2, 3.4, 1.4, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [5.4, 3.4, 1.5, 0.4],
       [5.2, 4.1, 1.5, 0.1],
       [5.5, 4.2, 1.4, 0.2],
       [4.9, 3

11. A partir dos dados lidos na matriz iris_2d, filtre as linhas que possuem (coluna 3) > 1.5 e (coluna 1) < 5.0.

12. A partir dos dados lidos na matriz iris_2d, calcule a **média** e **desvio padrão** de cada coluna presente nos dados. Imprima o resultado.

13. Crie um repositório no Github e versione todos os códigos que desenvolveram nesta lista de exercícios.